compiler/sem/sigmatch.nim

```diff
@@ -1733,15 +1728,35 @@ typeRel can be used to establish various relationships between types:
         effectiveArgType = a.skipTypes({tyRange, tyGenericInst,
                                         tyBuiltInTypeClass, tyAlias,
                                         tySink})
-        typeClassMatches = targetKind == effectiveArgType.kind and
-                             not effectiveArgType.isEmptyContainer
-      if typeClassMatches or
-        (targetKind in {tyProc, tyPointer} and effectiveArgType.kind == tyNil):
-        if doBind:
-          put(c, f, a)
-        return isGeneric
+
+      if targetKind == effectiveArgType.kind and
+         not effectiveArgType.isEmptyContainer:
+        # same type kind -> match
+        result = isGeneric
       else:
-        return isNone
+        case targetKind
+        of tyOrdinal, tyOpenArray, tyVarargs:
+          # these type classes of these act the same as if using, for example,
+          # ``openArray[T]``. The generic type is already stored as the base
+          # type -- but don't bind anything
+          result = typeRel(c, f.base, a, {trDontBind})
+        of tyRange:
+          if a.kind == tyRange:
+            result = isGeneric
+        of tyArray:
+          # XXX: empty array types match for the array type class, but this is
+          #      inconsistent with, for example, sequence types
```

Comment:
further observation: In hindsight after my comment/observation with respect to bind-once behaviour of type classes, I think this behaviour makes a certain amount of sense, as an empty array with an unbound contained type (`any`) does very much still satisfy the type class.

Reply:
> clarifying for myself, this is where a literal `[]` would be typed as an "empty array" and that would match `array` the type class?

Yep.

> I think this behaviour makes a certain amount of sense

I also think it makes, although there is the problem that `proc p(x: array)` invoked as `p([])` would cause the compiler to crash. In addition, the behaviour is inconsistent with `seq`: a `@[]` (`(Sequence (Empty))`) doesn't match the `seq` type class.

In any way, I just noticed that `tySet` also needs to be special-cased in the same way that `tyArray` is.